src/main/java/net/minecraftforge/eventbus/LockHelper.java

```diff
@@ -12,46 +13,48 @@
  * yet still manages to properly deal with many threads.
  */
 public class LockHelper<K,V> {
-    private ReadWriteLock lock = new ReentrantReadWriteLock(true);
-    private final Map<K, V> map;
+    private final IntFunction<Map<K, V>> mapConstructor;
+    /**
+     * Only modify this map while holding the lock object!
+     */
+    private final Map<K, V> backingMap;
+    @Nullable
+    private volatile Map<K, V> readOnlyView = null;
+    private Object lock = new Object();
 
-    public LockHelper(Map<K, V> map) {
-        this.map = map;
+    public LockHelper(IntFunction<Map<K, V>> mapConstructor) {
+        this.mapConstructor = mapConstructor;
+        this.backingMap = mapConstructor.apply(0);
```

Comment:
Maybe use a bigger initial capacity as it's unlikely no events will exist?

Reply:
done